### PR TITLE
feat(hook-store): Add balance diff on hook context

### PR DESCRIPTION
**Pull request by yvesfracari**
# Summary

This PR adds the balance diff information from the bff simulated bundle endpoint with the swap widget. It also sends the current balance diff to the hook app.

# To Test

Swap widget can get token balance changes on pre-hooks.
1. Run one pre hook that will change your tokens (I recommend the Airdrop hook).
2. Check that your balance of the withdraw hook changed (probably this hook will not be on token list you have to import it) on the swap widget components

Swap widget isn't triggering simulations when there is no hook or in other swap types (limit, swap, etc).
1. Go to one tab, play around and check if any `simulateBundle` post request is triggered

Hook Apps can access the current balance diff.
1. Run the app locally
2. Open the build hook on a scenario where there should be a balance diff informed for the hook.
3. Check the console if the balance differences are correct.

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    return mergeBalanceDiffs(preHookBalanceDiff, orderMockBalanceDiff)
+  }, [preHookBalanceDiff, orderMockBalanceDiff])
+
+  const postHookBalanceDiff = useMemo(() => {
+    // is adding the first post hook or simulation not available
+    if (!data || !postHooks) return firstPostHookBalanceDiff
+
+    const lastPostHook = postHooks[postHooks.length - 1]
+    return data[lastPostHook?.uuid]?.cumulativeBalancesDiff || firstPostHookBalanceDiff
+  }, [data, postHooks, orderMockBalanceDiff, preHookBalanceDiff])
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:56]: `firstPostHookBalanceDiff` is used in the hook but it is not part of the hook deps, while `orderMockBalanceDiff` and `preHookBalanceDiff` are but are not used in the hook.

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    if (!hookToEditIndex && isPreHook) return EMPTY_BALANCE_DIFF
+
+    // is editing first postHook -> return
+    if (!hookToEditIndex && !isPreHook) return firstPostHookBalanceDiff
+
+    // is editing a non first hook, return the latest available hook state
+    const previousHookIndex = hookToEditIndex - 1
+
+    return data[otherHooks[previousHookIndex]?.uuid]?.cumulativeBalancesDiff || EMPTY_BALANCE_DIFF
+  }, [data, hookToEditDetails])
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:75]: Also missing dependencies, like `isPrehook`, `preHooks`, `postHooks`, `firstPostHookBalanceDiff`

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  return useMemo(() => {
+    if (hookToEditDetails?.uuid) return hookToEditBalanceDiff
+    if (isPreHook) return preHookBalanceDiff
+    return postHookBalanceDiff
+  }, [data, orderParams, preHooks, postHooks])
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:81]: Also missing dependencies.
I know, the linter not working hurts a lot.

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  second: Record<string, Record<string, string>>,
+): Record<string, Record<string, string>> {
+  const result: Record<string, Record<string, string>> = {}
+
+  // Helper function to add BigNumber strings
+  const addBigNumberStrings = (a: string, b: string): string => {
+    const bigA = BigNumber.from(a)
+    const bigB = BigNumber.from(b)
+    return bigA.add(bigB).toString()
+  }
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useBalancesDiff.ts:95]: Please move nested fn out, it doesn't need to be declared on every call. Or does it?

File: apps/cowswap-frontend/src/modules/swap/containers/SwapWidget/index.tsx [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import { ReactNode, useCallback, useMemo, useState } from 'react'
 
-import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
+// import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/containers/SwapWidget/index.tsx:3]: Remove?

File: apps/cowswap-frontend/src/modules/swap/hooks/useSwapButtonContext.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import { useMemo } from 'react'
 
-import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
+// import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/hooks/useSwapButtonContext.ts:3]: Same

File: apps/cowswap-frontend/src/modules/swap/hooks/useSwapState.tsx [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { useCallback, useMemo } from 'react'
 
-import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
+// import { useCurrencyAmountBalance } from '@cowprotocol/balances-and-allowances'
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/hooks/useSwapState.tsx:3]: Same

File: apps/cowswap-frontend/src/modules/tokensList/containers/SelectTokenWidget/index.tsx [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import { useCallback, useState } from 'react'
 
-import { useTokensBalances } from '@cowprotocol/balances-and-allowances'
+// import { useTokensBalances } from '@cowprotocol/balances-and-allowances'
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/tokensList/containers/SelectTokenWidget/index.tsx:3]: Same

**Comment by shoom3301** [comment]
https://github.com/cowprotocol/cowswap/pull/5039 - this PR already includes changed from the current one, so I think we can keep only 5039

**Review by elena-zh — APPROVED** [review]
Tested in #5039: balance differences feature works great besides 1 issue: 
1. Add a pre-hook that may cause a balance change
2. Pick this token in the Sell field
3. Go to the Swap/Limit/TWAP form
4. Press on the Max button on these forms --> increased balance will be picked on the form. 

It would be nice not to include the balance difference on these forms.

**Comment by yvesfracari** [comment]
These changes are included in this [PR](https://github.com/cowprotocol/cowswap/pull/5039), so I will be closing this PR as @shoom3301 suggested.

**Pull request closed without merging**